src/kernels/jupyter/launcher/serverUriStorage.ts

```diff
@@ -140,58 +142,56 @@ export class JupyterServerUriStorage implements IJupyterServerUriStorage {
             undefined
         );
     }
-    public getUri(): Promise<string> {
+    private getUris(): Promise<string[]> {
         if (!this.currentUriPromise) {
             this.currentUriPromise = this.getUriInternal();
         }
 
         return this.currentUriPromise;
     }
-    public async getRemoteUri(): Promise<string | undefined> {
-        const uri = await this.getUri();
-        switch (uri) {
-            case Settings.JupyterServerLocalLaunch:
-                return;
-            case Settings.JupyterServerRemoteLaunch:
-                // In `getUriInternal` its not possible for us to end up with Settings.JupyterServerRemoteLaunch.
-                // If we do, then this means the uri was never saved or not in encrypted store, hence no point returning and invalid entry.
-                return;
-            default:
-                return uri;
-        }
+    public async getRemoteUris(): Promise<string[]> {
+        return this.getUris();
     }
     public async setUriToLocal(): Promise<void> {
-        await this.setUri(Settings.JupyterServerLocalLaunch);
-    }
-    public async setUriToRemote(uri: string, displayName: string): Promise<void> {
-        await this.setUri(uri);
-        await this.addToUriList(uri, Date.now(), displayName);
+        this.remoteUris = [];
+        this.currentUriPromise = Promise.resolve(this.remoteUris);
+        await this.serverConnectionType.setIsLocalLaunch(true);
+        this._onDidChangeUri.fire();
     }
-
-    public async setUri(uri: string) {
+    public async addRemoteUri(uri: string, displayName: string): Promise<void> {
+        this.remoteUris.push(uri);
         // Set the URI as our current state
-        this.currentUriPromise = Promise.resolve(uri);
-        if (uri === Settings.JupyterServerLocalLaunch) {
-            await this.serverConnectionType.setIsLocalLaunch(true);
-        } else {
-            await this.addToUriList(uri, Date.now(), uri);
-            await this.serverConnectionType.setIsLocalLaunch(false);
+        this.currentUriPromise = Promise.resolve(this.remoteUris);
+        await this.addToUriList(uri, Date.now(), displayName);
+        await this.serverConnectionType.setIsLocalLaunch(false);
 
-            // Save in the storage (unique account per workspace)
-            const key = this.getUriAccountKey();
-            await this.encryptedStorage.store(Settings.JupyterServerRemoteLaunchService, key, uri);
-        }
+        // Save in the storage (unique account per workspace)
+        await this.encryptedStorage.store(
+            Settings.JupyterServerRemoteLaunchService,
+            this.getUrisAccountKey(),
+            JSON.stringify(this.remoteUris)
+        );
         this._onDidChangeUri.fire();
     }
-    private async getUriInternal(): Promise<string> {
+
+    private async getUriInternal(): Promise<string[]> {
```

Comment:
Looks like this could use a rename

Reply:
any suggestions.

Reply:
I was just thinking due to it being a plural now. getUrisInternal.

Reply:
Done